### PR TITLE
Add basic support for array lengths in types

### PR DESCRIPTION
**Pull request by lf-**
This recognizes `let a = [1u8, 2, 3]` as having type `[u8; 3]` instead
of the previous `[u8; _]`. Byte strings and `[0u8; 2]` kinds of range
array declarations are unsupported as before.

I don't know why a bunch of our rustc tests had single quotes inside
strings un-escaped by `UPDATE_EXPECT=1 cargo t`, but I don't think it's
bad? Maybe something in a nightly?

### CODE REVIEW

**Comment by lf-** [comment]
*the ci comes back* i see, those were indeed not to be updated, yet. well, fyi, some nightly will break these tests when you update ;-)

Should I revert my updates to those?

**Comment by flodiebold** [comment]
Good idea to start with array expressions, that sidesteps the whole const expression body lowering topic. I'll probably review this this evening.

**Comment by lf-** [comment]
> > I don't know why a bunch of our rustc tests had single quotes inside
> > strings un-escaped by `UPDATE_EXPECT=1 cargo t`
> 
> I think that means you're on an older Rust version

Quite the opposite, in fact :) Today's nightly. But I'm reverting it because it's not right for stable, which is nominally the target for the tests.

**Comment by lf-** [comment]
booped into Zulip for the traits wg: https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/Handling.20non.20evaluatable.20constants'.20equality/near/238386348

**Comment by lf-** [comment]
> LGTM. I'd be interested in what @nikomatsakis or @rust-lang/wg-traits think about handling unknown/unevaluatable consts though.
> 
> Also, if you haven't yet, would you mind running `rust-analyzer analysis-stats .` at least on the rust-analyzer repo itself to make sure it doesn't crash, and ideally compare the output before/after?

new:

```
Database loaded:     470.55ms, 256minstr
  crates: 34, mods: 661, decls: 12808, fns: 10109
Item Collection:     7.79s, 67ginstr
Inference:           17.38s, 103ginstr
Total:               25.17s, 170ginstr
```

old:

```
Database loaded:     621.49ms, 258minstr
  crates: 34, mods: 660, decls: 12804, fns: 10107
Item Collection:     7.84s, 67ginstr
5030/10107 49% processing: handlers::generate_function::tests::add_function_not_applicable_if_unresolved_variable_in_call_is_selecte
44/10107 49% processing: handlers::replace_if_let_with_match::tests::test_replace_if_let_with_match_doesnt_unwrap_multiline_expressi
5/10107 49% processing: handlers::replace_if_let_with_match::tests::test_replace_if_let_with_match_doesnt_unwrap_multiline_expressio
52/10107 49% processing: handlers::replace_if_let_with_match::tests::test_replace_match_with_if_let_doesnt_unwrap_multiline_expressi
3/10107 49% processing: handlers::replace_if_let_with_match::tests::test_replace_match_with_if_let_doesnt_unwrap_multiline_expressio
258/10107 52% processing: handlers::extract_struct_from_enum_variant::tests::test_extract_enum_not_applicable_for_element_with_no_fi
9/10107 52% processing: handlers::extract_struct_from_enum_variant::tests::test_extract_enum_not_applicable_for_element_with_no_fiel
320/10107 52% processing: handlers::expand_glob_import::tests::expanding_is_not_applicable_if_target_module_is_not_accessible_from_c
1/10107 52% processing: handlers::expand_glob_import::tests::expanding_is_not_applicable_if_target_module_is_not_accessible_from_cur
409/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_bad_cursor_clos
10/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_bad_cursor_closu
1/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_already_result_st
2/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_already_result_st
3/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_already_result_cl
4/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_return_type_already_result_cl
27/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_with_tail_block_like_match_r
8/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_with_tail_block_like_match_re
9/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_with_tail_block_like_match_de
30/10107 53% processing: handlers::wrap_return_type_in_result::tests::wrap_return_type_in_result_simple_with_tail_block_like_early_r
51/10107 53% processing: handlers::introduce_named_lifetime::tests::test_function_add_lifetime_to_params_in_presence_of_other_lifeti
2/10107 53% processing: handlers::introduce_named_lifetime::tests::test_function_add_lifetime_to_params_in_presence_of_other_lifetim
3/10107 53% processing: handlers::introduce_named_lifetime::tests::test_function_not_applicable_without_self_and_multiple_unnamed_pa
  exprs: 279717, ??ty: 2519 (0%), ?ty: 1814 (0%), !ty: 1081
Inference:           17.85s, 103ginstr
Total:               25.68s, 170ginstr
```

I'm going to consider that likely an insignificant difference, which does make a fair bit of sense given we are not doing much work to achieve this.

File: crates/hir_ty/src/consts.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//! Handling of concrete const values
+
+/// A concrete constant value
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ConstScalar {
+    // for now, we only support the trivial case of constant evaluating the length of an array
+    Usize(usize),
```
[the remark this answers is not in the document]
Comment (lf-) [reply crates/hir_ty/src/consts.rs]: changed.

**Comment by lf-** [comment]
I've addressed the actionable review feedback.

Currently I'm about halfway through doing https://github.com/rust-analyzer/rust-analyzer/pull/2834#discussion_r366196658, #540, etc, to be able to parse literals in RA, which we can use to add lengths to byte strings and to cheat on putting lengths on arrays with literal lengths. This work will be in a future PR based on this one for ease of review and I'll file that when  I get it done.

**Pull request merged**